### PR TITLE
[FSDP] Support Qwen3-30B-A3B and analyse memory usage

### LINKED ISSUE #1060: [FSDP] Model supporting Qwen3 30BA3B
Currently the Qwen3 30BA3B running in FSDP is not so stable, occupying much memory than expected, this part needs more serious analysis and may have some potential bug need to be fixed

### PR DESCRIPTION
**Pull request by mingMelody**
Given the constraints on my computing resources, I have adjusted `num_hidden_layers` to 6 for training purposes.

Regarding memory usage analysis, I plan to break down the analysis into the following components, with detailed insights to be added in commit comments. If there are other aspects that need to be analyzed, please let me know:

- [x] Model weights
- [x] Optimizer state
- [x] Activations

Closes #1060

### CODE REVIEW

**Comment by mingMelody** [comment]
## Model Weights Analysis

### Model Parameters and Expected Size

The table below outlines the major parameter blocks of the model and their expected size in `BF16` and `FP32` precision.

| Parameter Block                               | Matrix Shape               | BF16    | FP32   |
| ------------------------------------ | ------------------ | ------- | ------ |
| Embedding                            | [151936,2048]      | 593.5MB | 1187MB |
| Qwen3MoeAttention::q_proj            | [2048,4096]        | 16MB    | 32MB   |
| Qwen3MoeAttention::k_proj            | [2048,512]         | 2MB     | 4MB    |
| Qwen3MoeAttention::v_proj            | [2048,512]         | 2MB     | 4MB    |
| Qwen3MoeAttention::o_proj            | [4096,2048]        | 16MB    | 32MB   |
| Qwen3MoeSparseMoeBlock::gate_up_proj | [128, 2*768, 2048] | 768MB   | 1536MB |
| Qwen3MoeSparseMoeBlock::down_proj    | [128, 768, 2048]   | 384MB   | 768MB  |
| lm_head                              | [2048,151936]      | 593.5MB | 1187MB |

For a 6-layer implementation, the total expected weight size is:
- BF16 (Bfloat16): `8.12GB`
- FP32 (Float32): `16.24GB`

### Experimental Observation(Base on 2 * H100)

During the actual run, when observing the memory consumption before and after the first model offload operation, we found that the `allocated_GB` usage was approximately `8.13GB`. This observation strongly suggested that the model weights were being stored in FP32 precision.

```
[2025-12-15 15:13:01] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 64.91, 'used_GB': 14.42, 'allocated_GB': 8.13, 'reserved_GB': 11.9}
[2025-12-15 15:13:07] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 76.81, 'used_GB': 2.52, 'allocated_GB': 0.0, 'reserved_GB': 0.0}
[2025-12-15 15:13:22] memory_utils.py:63 - [Rank 0] Memory-Usage after wake_up model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 67.42, 'used_GB': 11.91, 'allocated_GB': 8.12, 'reserved_GB': 8.86}
[2025-12-15 15:13:48] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 8.5, 'used_GB': 70.82, 'allocated_GB': 24.44, 'reserved_GB': 67.64}
[2025-12-15 15:14:04] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 75.52, 'used_GB': 3.81, 'allocated_GB': 0.06, 'reserved_GB': 0.58}
[2025-12-15 15:14:10] memory_utils.py:63 - [Rank 0] Memory-Usage after wake_up model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 49.52, 'used_GB': 29.8, 'allocated_GB': 24.44, 'reserved_GB': 26.58}
[2025-12-15 15:14:12] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 24.04, 'used_GB': 55.28, 'allocated_GB': 24.44, 'reserved_GB': 52.06}
[2025-12-15 15:14:15] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 75.52, 'used_GB': 3.81, 'allocated_GB': 0.06, 'reserved_GB': 0.58}
[2025-12-15 15:14:25] memory_utils.py:63 - [Rank 0] Memory-Usage after wake_up model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 49.52, 'used_GB': 29.8, 'allocated_GB': 24.44, 'reserved_GB': 26.58}
[2025-12-15 15:14:27] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 8.67, 'used_GB': 70.66, 'allocated_GB': 24.44, 'reserved_GB': 67.43}
[2025-12-15 15:14:30] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.33, 'free_GB': 75.52, 'used_GB': 3.81, 'allocated_GB': 0.06, 'reserved_GB': 0.58}

```

To confirm the storage precision, the model loading  [code](https://github.com/THUDM/slime/blob/main/slime/backends/fsdp_utils/actor.py#L92-L96)  was explicitly modified to force BF16 precision using dtype=torch.bfloat16:
```
model = self.get_model_cls().from_pretrained(
          self.args.hf_checkpoint,
          trust_remote_code=True,
          attn_implementation=self.args.attn_implementation,
          dtype=torch.bfloat16
      )
```

The memory consumption was reduced by approximately half (`8.13GB` $\rightarrow$ `4.17GB`), definitively confirming that the model weights were initially stored in `FP32`
```
[2025-12-16 11:52:05] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 71.33, 'used_GB': 8.1, 'allocated_GB': 4.17, 'reserved_GB': 5.63}
[2025-12-16 11:52:09] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 76.96, 'used_GB': 2.48, 'allocated_GB': 0.0, 'reserved_GB': 0.0}
[2025-12-16 11:52:14] memory_utils.py:63 - [Rank 0] Memory-Usage after wake_up model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 72.26, 'used_GB': 7.18, 'allocated_GB': 4.15, 'reserved_GB': 4.17}
[2025-12-16 11:52:42] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 47.19, 'used_GB': 32.24, 'allocated_GB': 12.36, 'reserved_GB': 29.11}
[2025-12-16 11:52:51] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 75.7, 'used_GB': 3.74, 'allocated_GB': 0.06, 'reserved_GB': 0.58}
[2025-12-16 11:52:56] memory_utils.py:63 - [Rank 0] Memory-Usage after wake_up model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 63.75, 'used_GB': 15.68, 'allocated_GB': 12.51, 'reserved_GB': 12.53}
[2025-12-16 11:52:57] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 40.73, 'used_GB': 38.7, 'allocated_GB': 12.51, 'reserved_GB': 35.54}
[2025-12-16 11:52:59] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 75.7, 'used_GB': 3.74, 'allocated_GB': 0.06, 'reserved_GB': 0.58}
[2025-12-16 11:53:04] memory_utils.py:63 - [Rank 0] Memory-Usage after wake_up model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 63.75, 'used_GB': 15.68, 'allocated_GB': 12.51, 'reserved_GB': 12.53}
[2025-12-16 11:53:05] memory_utils.py:63 - [Rank 0] Memory-Usage before offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 40.31, 'used_GB': 39.13, 'allocated_GB': 12.51, 'reserved_GB': 35.97}
[2025-12-16 11:53:07] memory_utils.py:63 - [Rank 0] Memory-Usage after offload model: {'gpu': '0', 'total_GB': 79.44, 'free_GB': 75.7, 'used_GB': 3.74, 'allocated_GB': 0.06, 'reserved_GB': 0.58}
```

This confirms that in the initial run, the model weights were being stored in `FP32` precision on each GPU, resulting in an `8.13GB`allocation (half of the total `FP32` size).


We set `param_dtype = torch.bfloat16` for FSDP, so why are we still seeing `FP32` being used? After checking the [pytorch documentation](https://docs.pytorch.org/docs/stable/distributed.fsdp.fully_shard.html#torch.distributed.fsdp.MixedPrecisionPolicy) and verifying with logging, it was found that the weights only adopt `BF16` during the computation process.
```
[36m(FSDPTrainRayActor pid=294696)[0m final_hidden_states dtype: torch.bfloat16, router_logits dtype: torch.bfloat16
[36m(FSDPTrainRayActor pid=294696)[0m expert weight dtype: torch.bfloat16
```

**Comment by mingMelody** [comment]
## Optimizer state Analysis

### Optimizer state Expected Size

For the Adam optimizer, the memory consumption is approximately $4\times$ the size of the model parameters. The table below outlines the expected memory occupation for the optimizer state.

| dtype | memory  |
| ----- | ------- |
| BF16  | 16.24GB |
| PF32  | 32.48GB |

### Experimental Observation(Base on 2 * H100)

The experimental data is consistent with the Model Weights Analysis.

We determined the memory usage of the optimizer state by calculating the difference in `allocated_GB` between the first and second iterations of the `Memory-Usage before offload model` measurement.

In the `FP32` configuration, the optimizer state occupied approximately `16 GB`. 
In the `BF16` configuration, the state occupied approximately `8 GB`.

These results align perfectly with the 2-GPU setup, indicating that the optimizer state is correctly sharded across both cards. The observation that setting the parameter `dtype` during the model loading phase successfully dictates the precision of the optimizer state confirms expected behavior. However, this tight coupling emphasizes the critical need for careful and deliberate setting of the `dtype`, as it directly and significantly influences total memory utilization and potential training stability. 

I still haven’t figured out how to set them up separately, so don’t make any changes yet.

**Comment by zju-stu-lizheng** [comment]
I encountered a similar issue while running Qwen3-30B MoE with Slime + FSDP:  
If I don’t explicitly specify the `dtype`, the allocated memory is significantly larger than when explicitly setting `dtype=bfloat16`.  

Is manual specification the only solution?

**Comment by mingMelody** [comment]
> I encountered a similar issue while running Qwen3-30B MoE with Slime + FSDP: If I don’t explicitly specify the `dtype`, the allocated memory is significantly larger than when explicitly setting `dtype=bfloat16`.
> 
> Is manual specification the only solution?

I haven't found a better solution for now. Furthermore, I'm not sure if using `BF16` for the optimizer state is reliable for training. This may require the implementation of mixed-precision training. You can also try using `PYTORCH_CUDA_ALLOC_CONF=expandable_segments:True` for training, but it seems to cause illegal memory access errors on certain drivers, which has prevented me from conducting effective tests on it.

**Comment by KlZOnmyway** [comment]
Following the same experimental setting as above (i.e., bfloat16, num_layers=6), after adding several print_memory() calls during the training step, we plotted the evolution of `used_GB`, `reserved_GB`, and `allocated_GB`.

In `memory_rank0_no_clean`, the continuous growth of `used_GB` is primarily driven by the accumulation of reserved memory. PyTorch’s caching allocator retains large chunks of memory obtained during peak phases (e.g., log-prob computation and backward) for reuse, causing the driver-visible used memory to monotonically increase.

This can lead to OOM even when `allocated_GB` remains moderate: once `reserved_GB` occupies most of the GPU memory, any subsequent higher peak (e.g., larger batch size, longer sequence length, or additional temporary workspace) may fail to allocate.

Calling `torch.cuda.empty_cache()` at every step significantly lowers the `used_GB` baseline by releasing cached blocks back to the CUDA driver; however, this comes at a noticeable performance cost due to frequent GPU memory deallocation and reallocation.

A practical workaround is to reduce peak memory usage, for example by using chunked log_softmax / entropy computation, and calling `empty_cache()` only after the loss computation and backward pass.
mem_compare

**Comment by mingMelody** [comment]
**FYI:**
**Memory Profiling:** Due to the issue in PyTorch ([Issue #170137](https://github.com/pytorch/pytorch/issues/170137)), the torch memory profiler is currently incompatible.
**KV Cache Impact**: Although KV caching was initially enabled during the training phase, subsequent tests showed that disabling it had no measurable impact on the memory occupation. You can disable the kv cache usage by setting `use_cache=False` (https://github.com/THUDM/slime/blob/main/slime/backends/fsdp_utils/actor.py#L888C9-L892C10)
**Activation Analysis**：you can disable the `--gradient-checkpointing` to analyze the full activation footprint.

**Comment by mingMelody** [comment]
@zhaochenyang20 Please review it. Let me know if I’ve missed any other critical memory bottlenecks.

**Comment by zhuzilin** [comment]
Thank you for the script and the memory usage. We currently observed severe bugs and performance issue with fsdp backend and decide to write it. Therefore I'll close the pr.
